src/main.cpp

```diff
@@ -638,22 +638,105 @@ unsigned int LimitOrphanTxSize(unsigned int nMaxOrphans) EXCLUSIVE_LOCKS_REQUIRE
     return nEvicted;
 }
 
-bool IsFinalTx(const CTransaction &tx, int nBlockHeight, int64_t nBlockTime)
+int64_t LockTime(const CTransaction &tx, int flags, const CCoinsView* pCoinsView, int nBlockHeight, int64_t nBlockTime)
 {
-    if (tx.nLockTime == 0)
-        return true;
-    if ((int64_t)tx.nLockTime < ((int64_t)tx.nLockTime < LOCKTIME_THRESHOLD ? (int64_t)nBlockHeight : nBlockTime))
-        return true;
-    BOOST_FOREACH(const CTxIn& txin, tx.vin)
-        if (!txin.IsFinal())
-            return false;
-    return true;
+    CCoins coins;
+
+    bool fEnforceBIP68 = tx.nVersion >= 2
+                      && flags & LOCKTIME_VERIFY_SEQUENCE;
+
+    // Will be set to the equivalent height- and time-based nLockTime
+    // values that would be necessary to satisfy all relative lock-
+    // time constraints given our view of block chain history.
+    int nMinHeight = 0;
+    int64_t nMinTime = 0;
+    // Will remain equal to true if all inputs are finalized
+    // (CTxIn::SEQUENCE_FINAL).
+    bool fFinalized = true;
+
+    BOOST_FOREACH(const CTxIn& txin, tx.vin) {
+        // Set a flag if we witness an input that isn't finalized.
+        if (CTxIn::SEQUENCE_FINAL == txin.nSequence)
+            continue;
+        else
+            fFinalized = false;
+
+        // Do not enforce sequence numbers as a relative lock time
+        // unless we have been instructed to.
+        if (!fEnforceBIP68)
+            continue;
+
+        // Sequence numbers equal to or above the locktime threshold
+        // are not treated as relative lock-times, nor are they given
+        // any consensus-enforced meaning at this point.
+        if (txin.nSequence >= CTxIn::SEQUENCE_LOCKTIME_THRESHOLD)
+            continue;
+
+        // Skip this input if it is not in the UTXO set. This should
+        // only ever happen in non-consensus code.
+        if (!pCoinsView->GetCoins(txin.prevout.hash, coins))
+            continue;
```

Comment:
"I believe this is incorrect, as LockTime() is called for the coinbase transaction by ContextualCheckBlock()" - @petertodd from [other branch](https://github.com/maaku/bitcoin/commit/d454248ad6eab502762ed0b0aaa37d0f646f1844#commitcomment-13327523)